src/tty/windows/conpty.rs

```diff
@@ -0,0 +1,260 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::{Pty, HANDLE, process_should_exit};
+
+use std::env;
+use std::mem;
+use std::ptr;
+use std::os::windows::io::IntoRawHandle;
+use std::sync::Arc;
+
+use dunce::canonicalize;
+use miow;
+use mio_anonymous_pipes::{AsyncAnonRead, AsyncAnonWrite};
+use winapi::shared::basetsd::{PSIZE_T, SIZE_T};
+use winapi::shared::ntdef::{HANDLE, HRESULT, LPCWSTR, LPWSTR};
+use winapi::shared::minwindef::{BYTE, DWORD};
+use winapi::shared::winerror::S_OK;
+use winapi::um::libloaderapi::{GetModuleHandleA, GetProcAddress};
+use winapi::um::processthreadsapi::{
+    CreateProcessW,
+    InitializeProcThreadAttributeList,
+    UpdateProcThreadAttribute,
+    STARTUPINFOW,
+    PROCESS_INFORMATION
+};
+use winapi::um::winbase::{STARTUPINFOEXW, EXTENDED_STARTUPINFO_PRESENT};
+use winapi::um::wincon::{COORD, HPCON};
+use widestring::U16CString;
+
+use config::{Config, Shell};
+use display::OnResize;
+use cli::Options;
+use term::SizeInfo;
+
+#[allow(non_snake_case)]
+struct ConptyApi {
+    CreatePseudoConsole: unsafe extern "system" fn (
+        COORD,
+        HANDLE,
+        HANDLE,
+        DWORD,
+        *mut HPCON
+    ) -> HRESULT,
+    ResizePseudoConsole: unsafe extern "system" fn (HPCON, COORD) -> HRESULT,
+    ClosePseudoConsole: unsafe extern "system" fn (HPCON) -> HRESULT
+}
+
+lazy_static! {
+    static ref CONPTY_API: Option<ConptyApi> = unsafe {
+        let hmodule = GetModuleHandleA("kernel32\0".as_ptr() as _);
+        assert!(!hmodule.is_null());
+
+        let cpc = GetProcAddress(hmodule, "CreatePseudoConsole\0".as_ptr() as _);
+        let rpc = GetProcAddress(hmodule, "ResizePseudoConsole\0".as_ptr() as _);
+        let clpc = GetProcAddress(hmodule, "ClosePseudoConsole\0".as_ptr() as _);
+
+
+        if cpc.is_null() || rpc.is_null() || clpc.is_null() {
+            None
+        } else {
+            Some(ConptyApi {
+                CreatePseudoConsole: mem::transmute(cpc),
+                ResizePseudoConsole: mem::transmute(rpc),
+                ClosePseudoConsole: mem::transmute(clpc)
+            })
+        }
+    };
+}
+
+pub struct Conpty {
+    pub handle: HPCON
+}
+
+impl Drop for Conpty {
+    fn drop(&mut self) {
+        // The pseusdoconsole might already have been closed by the console process
+        // exiting. If that's not the case, close it here.
+        if (!process_should_exit()) {
+            let api = CONPTY_API.as_ref().unwrap();
+            let result = unsafe { (api.ClosePseudoConsole)(self.handle) };
+
+            // For future reference, if the pseudoconsole is already closed
+            // then result == 1 (as of testing in Nov 2018).
```

Comment:
Was this comment left over from using a raw 1? I think the `S_OK` constant probably makes this obvious?

Reply:
This should still be resolved.

Reply:
The opposite intention - S_OK constant maps to zero. The intention of the comment is that if the pseudoconsole is already closed (e.g. because the user typed `exit` to close the console process), then ClosePseudoConsole fails with the error code 1.

The `!process_should_exit()` condition prevents this failure from occuring as the code is currently structured. As I couldn't find a description of the error case in the docs I wanted to document it here incase future innocent refactoring leads to the assert getting tripped. In which case I wanted the future dev to know what might be causing that.

I'll reword the comment a little to hopefully make it clearer.